velox/functions/prestosql/GreatestLeast.cpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions {
+
+namespace {
+
+template <bool>
+class ExtremeValueFunction;
+
+using LeastFunction = ExtremeValueFunction<true>;
+using GreatestFunction = ExtremeValueFunction<false>;
+
+/**
+ * This class implements two functions:
+ *
+ * greatest(value1, value2, ..., valueN) → [same as input]
+ * Returns the largest of the provided values.
+ *
+ * least(value1, value2, ..., valueN) → [same as input]
+ * Returns the smallest of the provided values.
+ **/
+template <bool isLeast>
+class ExtremeValueFunction : public exec::VectorFunction {
+ private:
+  template <typename T>
+  bool shouldOverride(const T& currentValue, const T& candidateValue) const {
+    return isLeast ? candidateValue < currentValue
+                   : candidateValue > currentValue;
+  }
+
+  template <typename T>
+  void applyTyped(
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    BaseVector::ensureWritable(rows, outputType, context->pool(), result);
+    BufferPtr resultValues =
+        (*result)->as<FlatVector<T>>()->mutableValues(rows.end());
+    T* __restrict rawResult = resultValues->asMutable<T>();
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+
+    rows.applyToSelected([&](int row) {
+      T currentValue = decodedArgs.at(0)->valueAt<T>(row);
+      for (auto i = 1; i < args.size(); ++i) {
+        auto candidateValue = decodedArgs.at(i)->template valueAt<T>(row);
+        if (shouldOverride(currentValue, candidateValue)) {
+          currentValue = candidateValue;
+        }
+      }
+      rawResult[row] = currentValue;
```

Comment:
In case of strings, we need to make sure to add a reference to string buffers to result.